File: data_structure_python/sort/insert_sort.py
class Sort:
    def __init__(self, array):
        self.arr = array

    def insert_sort(self):
        n = len(self.arr)
        for i in range(n):
            temp = self.arr[i]
            j = i
            while j > 0 and self.arr[j - 1] > temp:
                self.arr[j] = self.arr[j - 1]
                j -= 1
            self.arr[j] = temp
        return self.arr

    def insert_sort_reverse(self):
        n = len(self.arr)
        for i in range(n):
            temp = self.arr[i]
            j = i
            while j > 0 and self.arr[j - 1] < temp:
                self.arr[j] = self.arr[j - 1]
                j -= 1
            self.arr[j] = temp
        return self.arr

    def __str__(self):
        return str(self.arr)


if __name__ == '__main__':
    arr = [6, 8, 3, 1, 3, 2]
    s = Sort(arr)
    s.insert_sort()
    print(s)
    s.insert_sort_reverse()
    print(s)

File: DataStructureJava/sorting/InsertSort.java
import java.util.Arrays;

public class InsertSort {
    private InsertSort(){}
    public static <E extends Comparable<E>> void sort(E[] arr) {
        for (int i = 0; i < arr.length; i++) {
            E temp = arr[i];
            int j;
            for (j = i; j - 1 >= 0 && temp.compareTo(arr[j - 1]) < 0; j--) {
                arr[j] = arr[j - 1]; //move every element 
            }
            arr[j] = temp; //insert into the right index
        }
    }


    public static void main(String[] args){

        int[] dataSize = {10000,100000};
        for (int n:dataSize){
            System.out.println(" Random array: ");

            Integer[] arr = ArrayGenerator.generatorRandomArray(n,n);
            Integer[] arr2 = Arrays.copyOf(arr,arr.length);
            SortingHelper.sortTest("InsertSort",arr);
            SortingHelper.sortTest("SelectSort",arr2);

            System.out.println();

            System.out.println(" Ordered Array: ");
            arr = ArrayGenerator.generateOrderArray(n);
            arr2 = Arrays.copyOf(arr,arr.length);
            SortingHelper.sortTest("InsertSort",arr);
            SortingHelper.sortTest("SelectSort",arr);
        }
    }
}
